File: python/luhn/luhn.py
import re


class Luhn:
    regex = r"[^0-9\s]"

    def __init__(self, card_number):
        self.card_number = card_number.strip()

    def _clean_card(self):
        return re.sub(self.regex, "", self.card_number).replace(" ", "")

    def _doesnt_have_valid_length(self):
        return len(self.card_number) <= 1

    def _doesnt_have_valid_chars(self):
        return len(re.findall(self.regex, self.card_number)) > 0

    def _calculate(self):
        card_number = self._clean_card()
        positions = [int(i) for i in range(len(card_number) - 2, -1, -2)]
        if len(positions) == 0:
            positions.append(0)

        for_calc = [int(char) for char in card_number]
        for position in positions:
            product = for_calc[position] * 2
            for_calc[position] = product if product < 9 else product - 9

        return sum(for_calc) % 10 == 0

    def valid(self):
        if self._doesnt_have_valid_length() or self._doesnt_have_valid_chars():
            return False
        return self._calculate()

File: go/luhn/luhn.go
package luhn

import (
	"regexp"
	"strconv"
	"strings"
)

func Valid(id string) bool {
	id = strings.TrimSpace(id)

	if id == "0" ||
		doesntHaveValidLenght(id) ||
		doesntHaveValidChars(id) {
		return false
	}

	return calculate(id)

}

func onlyNumbers(id string) []string {
	re := regexp.MustCompile(`\D`)
	return strings.Split(re.ReplaceAllString(id, ""), "")
}

func doesntHaveValidLenght(id string) bool {
	return len(id) <= 1
}

func doesntHaveValidChars(id string) bool {
	re := regexp.MustCompile(`[^\d\s]`)
	return re.FindAllString(id, -1) != nil
}

func calculate(id string) bool {
	var positions, forCalc []int
	var total int
	card := onlyNumbers(id)
	positions = getPositions(card)

	for _, number := range card {
		number, _ := strconv.Atoi(number)
		forCalc = append(forCalc, number)
	}

	for _, position := range positions {
		forCalc[position] *= 2
		if forCalc[position] > 9 {
			forCalc[position] -= 9
		}
	}

	for _, number := range forCalc {
		total += number
	}

	return total%10 == 0
}

func getPositions(card []string) []int {
	var positions []int
	i := len(card) - 2
	for {
		if i <= -1 {
			break
		}
		positions = append(positions, i)
		i -= 2
	}
	return positions
}
